agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,319 @@
+	defer h.logResources("Consumed resources after task consume call", taskArn)
+
+	// Check if already consumed
+	_, ok := h.taskConsumed[taskArn]
+	if ok {
+		// Nothing to do, already consumed, return
+		logger.Info("Resources pre-consumed, continue to task creation", logger.Fields{"taskArn": taskArn})
+		return true, nil
+	}
+
+	ok, err := h.consumable(resources)
+	if err != nil {
+		logger.Error("Resources failing to consume, error in task resources", logger.Fields{
+			"taskArn":   taskArn,
+			field.Error: err,
+		})
+		return false, err
+	}
+	if ok {
+		// CPU
+		cpu := *h.consumedResource[CPU].IntegerValue + *resources[CPU].IntegerValue
+		h.consumedResource[CPU].SetIntegerValue(cpu)
+
+		// MEM
+		mem := *h.consumedResource[MEMORY].IntegerValue + *resources[MEMORY].IntegerValue
+		h.consumedResource[MEMORY].SetIntegerValue(mem)
+
+		// PORTS
+		portsResource, ok := resources[PORTSTCP]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			for _, port := range taskPortsSlice {
+				// Create a copy to assign it back as "PORTS_TCP"
+				newPortResource := h.consumedResource[PORTSTCP]
+				newPorts := append(h.consumedResource[PORTSTCP].StringSetValue, port)
+				newPortResource.StringSetValue = newPorts
+				h.consumedResource[PORTSTCP] = newPortResource
+			}
+		}
+
+		// PORTS_UDP
+		portsResource, ok = resources[PORTSUDP]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			for _, port := range taskPortsSlice {
+				newPortResource := h.consumedResource[PORTSUDP]
+				newPorts := append(h.consumedResource[PORTSUDP].StringSetValue, port)
+				newPortResource.StringSetValue = newPorts
+				h.consumedResource[PORTSUDP] = newPortResource
+			}
+		}
+
+		// GPU
+		*h.consumedResource[GPU].IntegerValue += *resources[GPU].IntegerValue
+
+		// Set consumed status
+		h.taskConsumed[taskArn] = true
+		logger.Info("Resources successfully consumed, continue to task creation", logger.Fields{"taskArn": taskArn})
+		return true, nil
+	}
+	logger.Info("Resources not consumed, enough resources not available", logger.Fields{"taskArn": taskArn})
+	return false, nil
+}
+
+// Functions checkConsumableIntType and checkConsumableStringSetType to be called
+// only after checking for resource map health
+func (h *HostResourceManager) checkConsumableIntType(resourceName string, resources map[string]*ecs.Resource) bool {
+	resourceConsumableStatus := *(h.initialHostResource[resourceName].IntegerValue) >= *(h.consumedResource[resourceName].IntegerValue)+*(resources[resourceName].IntegerValue)
+	return resourceConsumableStatus
+}
+
+func (h *HostResourceManager) checkConsumableStringSetType(resourceName string, resources map[string]*ecs.Resource) bool {
+	resourceSlice := resources[resourceName].StringSetValue
+	// Check intersection of resource StringSetValue is empty with consumedResource
+	for _, obj1 := range resourceSlice {
+		for _, obj2 := range h.consumedResource[resourceName].StringSetValue {
```

Comment:
Is it worth optimizing this from O(mn) to O(m+n) using maps? I am imagining this taking 200K iterations on a really busy instance with 20K consumed ports and a new task that wants to consume 10 ports.

Reply:
Good suggestion, let me get that in.

Reply:
nit: Maybe we should create a map with `resourceSlice` values instead as its size should be considerably smaller than consumed ports size.